### ISSUE TITLE
data processing problem

### ISSUE THREAD

**Issue by BruceWeiii** [csyxwei/FFWM#3]
Hi, your work is very good!
I encountered some problems in the data processing. I don't understand how to get the masks folder and the landmarks.npy file. Could you tell me more details about the data processing?  And could you share your multipie dataset? Thank you very much, and I will look forward to your reply.

**Comment by csyxwei** [comment]
Hi, you can follow these steps to process your data:

### Get landmark JSON files

1. Put your training/testing images into one folder
2. Register your face plus plus account, and change the Key and API Secret in [get_landmarks.py](https://github.com/csyxwei/FFWM/blob/master/data_process/get_landmarks.py)

![image](https://user-images.githubusercontent.com/23421814/131942513-7728ec8a-a062-4221-beed-252f3df0a2a1.png)


![image](https://user-images.githubusercontent.com/23421814/131942538-75fecb35-b687-4f2c-aaee-3efe427ae6a2.png)

3. Change the path and suf_num in [get_landmarks.py](https://github.com/csyxwei/FFWM/blob/master/data_process/get_landmarks.py)

![image](https://user-images.githubusercontent.com/23421814/131942898-ec94015c-3621-43e6-80b9-daf5a83f43f7.png)

4. Run the get_landmarks.py, then the processed JSON files will be saved in the json_path

### Get facial masks

1. Clone the [face-parsing.PyTorch](https://github.com/zllrunning/face-parsing.PyTorch).
2. Follow the README to process your training/testing images, and save the processed mask images into one folder.
    -  When processing, you should firstly crop your image to a square shape (add additional process code before the `resize` step in [test.py](https://github.com/zllrunning/face-parsing.PyTorch/blob/d2e684cf1588b46145635e8fe7bcc29544e5537e/test.py#L70)), and after processing, you need to resize and pad mask image to its original resolution.
    - The labels we used to train the FFWM are `1-17`. Check the README in the [CelebAMask-HQ](https://github.com/switchablenorms/CelebAMask-HQ/tree/master/face_parsing) to see what these labels mean. In the final mask, the parts related to the label `1-17` are 1, and the other is 0.

### Align your training/testing images 

1. Uncomment the code (`Multi-PIE training set` part) in process.py
2. Change the path to yours.
![image](https://user-images.githubusercontent.com/23421814/131945638-931ae893-bfd6-495c-ae50-64ad2f99a183.png)
3. Run the process.py

Then the landmarks.py and processed images and masks are saved in the `save_dir`.   `MultiPIE testing images`  and `LFW testing images` processing codes are also provided.

You can get the MultiPIE from their [website](https://www.cs.cmu.edu/afs/cs/project/PIE/MultiPie/Multi-Pie/Home.html). If you want to process other datasets, you should:

1. Modify the `s2f` function. This function is used to get the frontal image name.

2. Change the image center coordinate in the `process_multipie_train` function.
![image](https://user-images.githubusercontent.com/23421814/131946373-f0e32ecf-93df-426b-ab99-f3a5a40f4679.png)

**Comment by BruceWeiii** [comment]
Hi，Thank you very much for your reply. When I run get_landmarks.py, I got HTTP Error 403: Forbidden. I have changed the Key and API Secret in get_landmarks.py. Will it have some impact on the result?

**Comment by csyxwei** [comment]
Hi, http Error 403: Forbidden is caused by frequent api requests, and our code will repeat the failed request until all images are processed. It will not impact the result.

**Comment by BruceWeiii** [comment]
Hi, Sorry to bother you again. I used multipie_crop_128 dataset to test. But I got pictures like this after running process.py.
input:  
![201_01_01_041_09_crop_128_img_S](https://user-images.githubusercontent.com/56987659/132807140-eb7dbda2-b82e-4ccc-ba05-e642d7547cfd.png)
target: 
![201_01_01_041_09_crop_128_img_F](https://user-images.githubusercontent.com/56987659/132807199-50e4d0a7-afbe-4719-83c5-5333ff5ff07a.png)
And the test result is:
 ![201_01_01_041_09_crop_128_fake_F128](https://user-images.githubusercontent.com/56987659/132807114-bf808bd5-0ece-49e2-9088-e818fe7affb5.png)

Is this a problem of data? Could you tell me how to fix it? Thank you very much. Looking forward to your reply.

**Comment by csyxwei** [comment]
Hi, the aligned image has something wrong. Could you provide one of your landmark json files and your process code?  These can help me find the problem.

**Comment by BruceWeiii** [comment]
Hi, I haven't the original multipie dataset image, and I used aligned 128x128 images shared by others on the Internet.
And here are my landmark json files and process code. 
[json and process file.zip](https://github.com/csyxwei/FFWM/files/7149156/json.and.process.file.zip)

**Comment by xsacha** [comment]
Is there a reason that the landmarks cannot just be uploaded rather than everyone re-generating it from Face++?

**Comment by csyxwei** [comment]
> Hi, I haven't the original multipie dataset image, and I used aligned 128x128 images shared by others on the Internet.
> And here are my landmark json files and process code.
> [json and process file.zip](https://github.com/csyxwei/FFWM/files/7149156/json.and.process.file.zip)

Hi, the black area in the image is caused by the different alignment method. In our alignment method, the final cropped face area is larger than your aligned 128x128 image contains, so it contains black area when you align it by our method. Using original multipie dataset to align can avoid this problem.

**Comment by csyxwei** [comment]
> Is there a reason that the landmarks cannot just be uploaded rather than everyone re-generating it from Face++?

Hi, it's my fault. I have uploaded the landmarks and masks to facilitate the data processing, you can download them from [GoogleDrive](https://drive.google.com/drive/folders/1U26FvuLtXraxPrRNxCxFka3DOvRuG4NW?usp=sharing) or [BaiduNetDisk](https://pan.baidu.com/s/1X62Atd9Q_USs0aGQk3WllA) (l98p).

**Comment by BruceWeiii** [comment]
Hi,  sorry to bother you again. My Multi-PIE dataset is incomplete. I don't know if you can share part of the dataset. If this is allowed, can you share the test set of Multi-PIE? Thank you very much, and I will look forward to your reply.

**Comment by csyxwei** [comment]
Sorry, according to the license, I can not share the dataset with you. You can also use your aligned dataset to prepare the `landmarks` and `masks` and train the ffwm with your dataset.

**Comment by ilycorn** [comment]
> Hi, http Error 403: Forbidden is caused by frequent api requests, and our code will repeat the failed request until all images are processed. It will not impact the result.

Hi, I also encountered the same problem. It has always been 403forbidden. I would like to ask how you solved it and whether you need to spend money to purchase permissions.
![image](https://user-images.githubusercontent.com/38346436/138407206-062f3720-1d9c-4aad-be08-a08203ed7051.png)

**Comment by csyxwei** [comment]
Hi, 

Face++ provides one free key for users and you don't need to spend money. 403 Forbidden is caused by frequent api requests, and our code will repeat the failed request until all images are processed. It may try many times to process one image successfully. 

You can check if there are processed landmarks in your `json_path `. If there is, you `key` and `skey` is right, just wait patiently. Or you can check your `key`, `skey` and `request url`.

**Comment by ilycorn** [comment]
Hello, can you give me an example of a successful code? I have tried many times without success. Thank you very much.

**Comment by csyxwei** [comment]
Hi,

I have run the code. Here is my result:

### 1. Replace the key and skey

![image](https://user-images.githubusercontent.com/23421814/139029512-8ae73547-4925-4eed-b6ba-8fc8f1d539a2.png)


### 2. Change the path

![image](https://user-images.githubusercontent.com/23421814/139029563-0046161c-44a5-4cce-b3a6-53bd3e16fbc7.png)

![image](https://user-images.githubusercontent.com/23421814/139029223-18d1c54a-9bf6-499a-93d8-72301fe9b804.png)


### 3. Run the code

![image](https://user-images.githubusercontent.com/23421814/139029355-582e0370-407c-4abb-9d8b-e8d1562ea9ca.png)

and the extracted landmarks are saved in `lm` folder.

![image](https://user-images.githubusercontent.com/23421814/139029405-16c0d23d-c001-48c8-b9ca-109b68da5491.png)

**Maybe you can check the http_utl in `face_plus_plus.py`, or key and skey in `get_landmarks.py`.  After that, if you also have the same error, you can contact me by email, I can share my code with you.**

![image](https://user-images.githubusercontent.com/23421814/139029985-f5fa55f1-1a7a-4671-9e4d-4b506007adf1.png)

**Comment by ilycorn** [comment]
Hello, my QQ number is 840549388.Can you add me?Thank you very much

**Comment by xsacha** [comment]
There's an easy solution to just upload the result images? So everyone doesn't have to go through this arduous process and computers aren't wasted generating the results 1000s of times.

**Comment by csyxwei** [comment]
> There's an easy solution to just upload the result images? So everyone doesn't have to go through this arduous process and computers aren't wasted generating the results 1000s of times.

Hi, 

For MultiPIE dataset, I have uploaded the landmarks and masks to facilitate the data processing. And for LFW dataset, I have uploaded the aligned images. One can download them from [GoogleDrive](https://drive.google.com/drive/folders/1U26FvuLtXraxPrRNxCxFka3DOvRuG4NW?usp=sharing) or [BaiduNetDisk](https://pan.baidu.com/s/1X62Atd9Q_USs0aGQk3WllA) (l98p). 

For customized dataset, you have to run the above steps to process your own data, it can not be skipped.
